yarnpkg/yarn
# Add support for config file option 'cafile'

**Summary**

With a private registry using a company CA issued certificate, yarn fails because nodejs does not use the OS SSL/TLS trust store (at least on Linux).  The same issue affects registries using self signed certificates and might also apply for the public registries in companies using SSL/TLS intercepting proxies.

This is change adds support for the `cafile`-option, which should point to a file path containing one or more PEM files concatenated together, like in `npm`.

I've chosen `cafile` and not the `ca` option, because it is simpler to point to the operating system TLS trust store, which often will already have the correct CA certificates for corporate computers.

This does not add support for disabling strict-ssl, because that will open for MITM attacks.

References #606, #631

**Test plan**

***Integration test***
I've added a test which sets up a web server with a self signed certificate and a trust store containing a normal CA and the self signed certificate, and verifies that a connection set up using RequestManager can perform a GET-request from that server.

***Manual test***
With default config an no cafile set, I get following stack trace:
```
~/yarn/bin/yarn install
yarn install v0.15.1
[1/4] Resolving packages...
[2/4] Fetching packages...
error unable to verify the first certificate
    at Error (native)
    at TLSSocket.<anonymous> (_tls_wrap.js:1017:38)
    at emitNone (events.js:67:13)
    at TLSSocket.emit (events.js:166:7)
    at TLSSocket._init.ssl.onclienthello.ssl.oncertcb.TLSSocket._finishInit (_tls_wrap.js:582:8)
    at TLSWrap.ssl.onclienthello.ssl.oncertcb.ssl.onnewsession.ssl.onhandshakedone (_tls_wrap.js:424:38)
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
```

With `cafile=/etc/pki/ca-trust/extracted/pem/tls-ca-bundle.pem` in `.npmrc` or `cafile /etc/pki/ca-trust/extracted/pem/tls-ca-bundle.pem` in `.yarnrc` installation succeeds
